File: DSA/Arshdeep-Kaur/Day 36/Day36_Question2.cpp
class Solution {
public:
    int dp1[100001][8];
    int dp2[100001][8];
    int make_top_equal(int i, int last, vector<int>& tops, vector<int>& bottom){
        if(i==tops.size()) {
            return 0;
        }

        if(i>0 and tops[i]!=last and bottom[i]!=last) {
            return 1e9;
        }

        if(dp1[i][last+1] != -1) {
            return dp1[i][last+1];
        }

        int op1 = INT_MAX, op2 = INT_MAX;

        if(last == -1 || bottom[i] == last) {
            op1 = 1 + make_top_equal(i+1,bottom[i],tops,bottom);
        }
        
        if(last == -1 || tops[i] == last) {
            op2 = make_top_equal(i+1,tops[i],tops,bottom);
        } 
        
        return dp1[i][last+1] = min(op1,op2);
    }

    int make_bottom_equal(int i, int last, vector<int>& tops, vector<int>& bottom){
        if(i==tops.size()) {
            return 0;
        }
        
        if(i>0 and tops[i]!=last and bottom[i]!=last) {
            return 1e9;
        }

        if(dp2[i][last+1] != -1) {
            return dp2[i][last+1];
        }
        
        int op1 = INT_MAX, op2 = INT_MAX;
        
        if(last == -1 || tops[i] == last){
            op1 = 1 + make_bottom_equal(i+1,tops[i],tops,bottom);
        }
        
        if(last == -1 || bottom[i] == last) {
            op2 = make_bottom_equal(i+1,bottom[i],tops,bottom);
        }
        
        return dp2[i][last+1] = min(op1,op2);
    }

    int minDominoRotations(vector<int>& tops, vector<int>& bottoms) {
        memset(dp1,-1,sizeof(dp1));
        memset(dp2,-1,sizeof(dp2));
        int ans =  min(make_top_equal(0,-1,tops,bottoms), make_bottom_equal(0,-1,tops,bottoms));
        return (ans >= 1e9) ? -1: ans;
    }
};

File: DSA/Manav-Rajput/Day36_question2.java
class Solution {
    public int minDominoRotations(int[] tops, int[] bottoms) {
        int n = tops.length;
        int rotations = check(tops[0], tops, bottoms, n);
        if (rotations != -1 || tops[0] == bottoms[0]) {
            return rotations;
        } else {
            return check(bottoms[0], tops, bottoms, n);
        }
    }
    private int check(int target, int[] tops, int[] bottoms, int n) {
        int topRotations = 0;
        int bottomRotations = 0;
        for (int i = 0; i < n; i++) {
            if (tops[i] != target && bottoms[i] != target) {
                return -1; 
            } else if (tops[i] != target) {
                topRotations++;
            } else if (bottoms[i] != target) {
                bottomRotations++;
            }
        }
        return Math.min(topRotations, bottomRotations);
    }
}
